---
title: "InterCellar User Guide"
author:
- name: Marta Interlandi
  affiliation: 
  - Institute of Medical Informatics, University of Muenster, Muenster (DE)
  email: marta.interlandi@uni-muenster.de
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('InterCellar')`"
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{InterCellar User Guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: InterCellar.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
    comment = "#>",
    error = FALSE,
    warning = FALSE,
    message = FALSE,
    crop = NULL
)
```

```{r, echo=FALSE, out.width='50%', fig.align='center'}
knitr::include_graphics(path = system.file("app", "www", "logo.png", 
                                           package="InterCellar", mustWork=TRUE))
```

# Introduction

`InterCellar` is a [Bioconductor](http://bioconductor.org) package that provides
an interactive Shiny app to enable the analysis of cell-cell interactions from
single-cell RNA sequencing (scRNA-seq) data. Every step of the analysis can be
performed interactively, thus not requiring any computational skills. Moreover,
`InterCellar` runs on your local machine, avoiding issues related to data privacy.

## Installation

`InterCellar` is currently under revision in [Bioconductor](https://www.bioconductor.org/).

You can install the development version from GitHub with:

```{r eval = FALSE}
# install.packages("remotes")
remotes::install_github("martaint/InterCellar", ref = "R4.0")
```

## Launching the app

Once InterCellar is successfully installed, it can be loaded as follow:

```{r setup}
library(InterCellar)
```

In order to start the app, please run the following command:

```{r demostart, eval=FALSE}
InterCellar::run_app( reproducible = TRUE)
```

`InterCellar` should be opening in a browser. If this does not happen automatically, 
please open a browser and navigate to the address shown (for example, `Listening on http://127.0.0.1:6134`). The flag `reproducible = TRUE` ensures that your results will
be reproducible across R sessions.

# Overview of the workflow

The following figure represents `InterCellar` analysis workflow. We could not miss the opportunity
of depicting it as an interstellar journey.

```{r, echo=FALSE, out.width='100%', fig.align='center'}
knitr::include_graphics(path = system.file("app", "www", "About_shiny.png", 
                                           package="InterCellar", mustWork=TRUE))
```

# Data upload

The first step of the workflow requires the upload of pre-computed results 
generated by an external tool capable of predicting cell-cell interactions. 
`InterCellar` supports both published tools such as [*CellPhoneDBv2*](https://www.cellphonedb.org/) 
[@efremova2020cellphonedb] or `r Biocpkg("SingleCellSignalR")` [@cabello2020singlecellsignalr], 
and custom results output of *ad hoc* methods.
For the purpose of this user guide, we will use *CellPhoneDB* results computed on 
a scRNA-seq dataset from [@tirosh2016dissecting]. 

By navigating to **1. Data** and **Upload**, we can simply select the folder containing
*CellPhoneDB* results from our local drive. `InterCellar` will read and pre-process 
the data and show the resulting table in **Table view**. 
The pre-processing step consists of:

* Mapping interaction pairs to the corresponding genes (hgnc symbols) 
* Annotating genes to their molecular function: L (ligand) or R (receptor)
* Re-ordering the interaction pairs listed as R-L to L-R


```{r, echo=FALSE, out.width='120%',fig.align='center'}
knitr::include_graphics(path = "screenshots/upload_cpdb.png")
```

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/table_view.png")
```


# Universes

Once the input data has been uploaded, `InterCellar` takes us to the exploration 
of three **Universes**. Each universe has its focus on a different conceptual group:
clusters, genes and functions. Specific filtering options can
be applied and multiple visualization choices are available to enable a 
deep exploration of the cellular communication.

## Cluster-verse

Focus of this universe are clusters of cells participating in the interactions.
The filtering options allow the user to fine tune the dataset by:

* excluding entire clusters from the data. All interactions related to the excluded
clusters will be excluded as well from further steps of the analysis;
* setting a minimum interaction score;
* (when available) changing the p-value threshold for significant interactions.

The analyst will be able to see the effect of these filtering steps by looking at the 
box showing the number of total interactions.

Moreover, three tabs are part of the Cluster-verse: **Network**, **Barplot** and **Table**.

The **Network** of clusters shows the overall cellular communication. Nodes represent 
different clusters while edges show the total number of paracrine interactions performed between 
two clusters. Edges that fall back on the same cluster represent autocrine interactions.

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/cl_verse_net.png")
```

The **Barplot** offers a clear representation of the total number of interactions per cluster,
divided in paracrine and autocrine interactions.


```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/cl_verse_bar.png")
```

In the **Table** panel, the analyst can restrict the data exploration to a specific focus,
by subsetting the data to one cluster of interest, called *viewpoint*, and one *flow* of communication among:

* Directed, outgoing interactions (L-R): for which the *viewpoint* cluster sends (expresses)
the ligand to the other clusters, that are in turn expressing the corresponding receptor;
* Directed, incoming interactions (R-L): for which the *viewpoint* expresses the receptor and awaits for the corresponding ligand sent by other clusters;
* Undirected interactions (L-L and R-R): for which both elements of an interaction pair are either ligands or receptors.


```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/cl_verse_table.png")
```


## Gene-verse

`InterCellar` second universe brings the exploration to the genes. Filtering options are available
when the input data has been generated by *CellPhoneDB* and allow the user to exclude interaction pairs (int-pairs) that have been annotated from unwanted sources.
The **Table** shows all unique int-pairs enriched in our data, regardless of the clusters in which these are found. `InterCellar` calculates a *uniqueness score* for each int-pair (from 0 to 1, with 1 being the most unique) related to the fraction of cluster pairs that are enriched by this int-pair. In other words, an int-pair that is found between all clusters will have uniqueness score close to 0, whereas an int-pair which is only found between e.g. clusters 1 and 2, will have uniqueness score close to 1, suggesting higher specificity of this int-pair to a small subset of cell populations. Other info included in this **Table** are [Ensembl](https://www.ensembl.org) and [UniProt](https://www.uniprot.org/) IDs of each gene, with hyperlinks to the respective web pages to facilitate a deeper analysis.


```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/g_verse_table.png")
```

Upon selection of one or multiple int-pairs from the previous **Table**, a dotplot is generated and visible in the **Dot Plot** panel. The analyst can decide to exclude unwanted clusters from the visualization as well as choose different colors for high and low int-pair score.

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/g_verse_dot.png")
```



## Function-verse

Landing in the **Function-verse**, the analyst is required to perform a functional annotation, before proceeding to the next steps of the exploration. To this scope, `InterCellar` offers multiple sources of functional annotations in terms of [Gene Ontology](http://geneontology.org) (queried from [Ensembl](http://www.ensembl.org), via the package [biomaRt](https://bioconductor.org/packages/biomaRt/)) and pre-downloaded pathways databases (from the package [graphite](https://bioconductor.org/packages/graphite/)). After selection of suitable sources, the annotation can be performed and a **Table** showing all functional terms annotated to each int-pair is displayed. Worth to note is the fact that a functional term is annotated to an int-pair only when the functional term is enriched in both genes, partners of the interaction. Thus, `InterCellar` approach to functional annotation is driven by the underlying concept of protein-protein interaction: a function may be activated only upon presence of both protein  partners (in our case, genes). This approach differs from the well-known methods for functional enrichment analysis (e.g. carried out by [ToppGene](https://toppgene.cchmc.org/enrichment.jsp)), that are not based on the concept of protein-protein interactions. 

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/f_verse_tab.png")
```

The **Barplot** panel summarizes the number of functional terms annotated for each source.

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/f_verse_barplot.png")
```

In the following panel, **Ranking**, functional terms are listed individually, along with information on *occurrence* (i.e. how many int-pairs have been annotated to this term) and *average uniqueness* (i.e. the average value of the uniqueness scores of each int-pair annotated to this term). 

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/f_verse_ranking.png")
```

By selecting one row of the **Ranking** table, we can explore the term of interest in the **Sunburst** plot. This visualization allows us to connect functions to int-pairs and clusters. On the left side of the panel, a table lists all int-pairs annotated to the term. On the right side, the sunburst plot is composed as follows:

* the selected functional term is shown in the inner circle of the plot;
* the inner ring shows all "first partner" clusters, enriched by the relevant int-pairs. Specifically, clusters on the inner ring express the first gene of each int-pair;
* the outer ring displays all "second partner" clusters, expressing the second gene of each int-pair;
* the width of each section represents the fraction of int-pairs found in that section (also shown when hovering on the inner ring sections);
* hovering on outer ring sections will show the individual int-pairs enriched in the cluster pairs.

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/f_verse_sunburst.png")
```


# Int-Pair Modules Analysis

The final step of InterCellar exploration allows the analyst to define and analyze **Int-Pair Modules**, i.e. groups of int-pairs that share a similar functional profile. To this aim, the choice of a *viewpoint* cluster and communication *flow* is required. InterCellar will subset the input data accordingly. This analysis can be repeated for each viewpoint and flow of interest. Don't forget to save plots and tables that you want to keep!

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/ipM_analysis.png")
```

To define the number of int-pair modules in the data subset, four visualizations are provided. On the left hand side, the optimal number of modules is calculated by InterCellar using (1) the elbow method on the total within sum of square and (2) the average silhouette width. Both methods are standard practice in cluster analysis and are supposed to serve as an aid to choose the optimal number of modules. However, the analyst is free to choose the number that best satisfies his/her needs, in terms of high (low) specificity of a module which is reflected in high (low) number of groups. For this purpose, the two visualization on the right hand side offer yet another way to investigate the optimal number of modules. A **dendrogram** of int-pairs shows the results of a hierarchical clustering obtained on the first two components of the **UMAP** underneath. Each point of the UMAP represents one int-pair (shown by hovering) and color-coding is consistent for both UMAP and dendrogram, equal to the number of modules chosen. Moreover, dendrogram and UMAP are initialized with the optimal number of modules chosen by the elbow method (giving usually higher resolution compared to the average silhouette).  

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/ipM_def.png")
```

Once the int-pair modules have been defined, InterCellar offers the possibility to visualize the int-pairs belonging to each module and the respective clusters in a **Circle plot**. Directed interactions are represented here by arrows originating from ligands (double segment) towards receptors (single segment). The **Table** panel summarizes the same info in a tabular format.

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/ipM_circle.png")
```

Last step of the int-pair modules analysis brings us back to functional terms. Upon selection of one int-pair module, InterCellar performs a permutation test to calculate empirical p-values assessing the significance of functional terms annotated to int-pairs of the module. A **Table**  displays functional terms that are found significant (p-value <= 0.05 by default) for the chosen int-pair module.

```{r, echo=FALSE, fig.align='center'}
knitr::include_graphics(path = "screenshots/ipM_function.png")
```


# References